crates/wasi-common/src/sys/stdio.rs

```diff
@@ -0,0 +1,94 @@
+use super::{fd, AsFile};
+use crate::handle::{Handle, HandleRights};
+use crate::sandboxed_tty_writer::SandboxedTTYWriter;
+use crate::wasi::types::{self, Filetype};
+use crate::wasi::{Errno, Result};
+use std::any::Any;
+use std::cell::Cell;
+use std::io::{self, Read, Write};
+
+pub(crate) trait StdioExt: Sized {
+    /// Create `Stdio` from `io::stdin`.
+    fn stdin() -> io::Result<Box<dyn Handle>>;
+    /// Create `Stdio` from `io::stdout`.
+    fn stdout() -> io::Result<Box<dyn Handle>>;
+    /// Create `Stdio` from `io::stderr`.
+    fn stderr() -> io::Result<Box<dyn Handle>>;
+}
+
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub(crate) enum Stdio {
+    In { rights: Cell<HandleRights> },
+    Out { rights: Cell<HandleRights> },
+    Err { rights: Cell<HandleRights> },
+}
+
+impl Handle for Stdio {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn try_clone(&self) -> io::Result<Box<dyn Handle>> {
+        Ok(Box::new(self.clone()))
+    }
+    fn get_file_type(&self) -> Filetype {
+        Filetype::CharacterDevice
+    }
+    fn get_rights(&self) -> HandleRights {
+        match self {
+            Self::In { rights } => rights.get(),
+            Self::Out { rights } => rights.get(),
+            Self::Err { rights } => rights.get(),
+        }
+    }
+    fn set_rights(&self, new_rights: HandleRights) {
+        match self {
+            Self::In { rights } => rights.set(new_rights),
+            Self::Out { rights } => rights.set(new_rights),
+            Self::Err { rights } => rights.set(new_rights),
+        }
+    }
+    // FdOps
+    fn fdstat_get(&self) -> Result<types::Fdflags> {
+        fd::fdstat_get(&self.as_file())
+    }
+    fn fdstat_set_flags(&self, fdflags: types::Fdflags) -> Result<()> {
+        if let Some(_) = fd::fdstat_set_flags(&self.as_file(), fdflags)? {
+            // OK, this means we should somehow update the underlying os handle,
+            // and we can't do that with `std::io::std{in, out, err}`, so we'll
+            // panic for now.
+            panic!("Tried updating Fdflags on Stdio handle by re-opening as file!");
+        }
+        Ok(())
+    }
+    fn read_vectored(&self, iovs: &mut [io::IoSliceMut]) -> Result<usize> {
+        let nread = match self {
+            Self::In { .. } => io::stdin().read_vectored(iovs)?,
+            _ => return Err(Errno::Badf),
+        };
+        Ok(nread)
+    }
+    fn write_vectored(&self, iovs: &[io::IoSlice]) -> Result<usize> {
+        let nwritten = match self {
+            Self::In { .. } => return Err(Errno::Badf),
+            Self::Out { .. } => {
+                // lock for the duration of the scope
+                let stdout = io::stdout();
+                let mut stdout = stdout.lock();
+                let nwritten = if self.is_tty() {
```

Comment:
Can you leave a comment specifically on why we have to handle the tty and non-tty cases separately? Whether `io::stdout()` gives us a tty or not depends on how the embedding process is invoked, right? Is there any way to avoid exposing that to the guest?

Reply:
Yeah, that's correct. I actually need to revisit at some point, but the reason we ended up with custom `Stdio` type which is a proxy for Rust's stdio handles, is because Windows was doing some crazy stdio redirects on the CI which were nontrivial to handle. I'll add some comments about for posterity. Thanks for spotting that one!

Reply:
Just clarify further, as far as the guest is concerned, `Stdio` should be transparent to them. As I mentioned above, the only reason we have a separate `Stdio` type in `wasi-common` is to correctly facilitate redirects on Windows. To elaborate further, in POSIX, we can get a stdio handle by opening a specific fd `{0,1,2}`. On Windows however, we need to issue a syscall that's separate from standard Windows "open" to get a console handle, and this is [`GetStdHandle`](https://docs.microsoft.com/en-us/windows/console/getstdhandle). This is exactly what Rust does and what is wrapped inside their `Stdio` object in the `libstd`. We wrap it here as well because of this nuance on Windows:

> The standard handles of a process may be redirected by a call to SetStdHandle, in which case GetStdHandle returns the redirected handle.

The [MSDN](https://docs.microsoft.com/en-us/windows/console/getstdhandle) also says this however:

> If the standard handles have been redirected, you can specify the CONIN$ value in a call to the CreateFile function to get a handle to a console's input buffer. Similarly, you can specify the CONOUT$ value to get a handle to a console's active screen buffer.

But I'm not that well versed in Windows programming to know how to do this properly, which is not to say we shouldn't (re-)investigate at some point.

Anyway, do you think this explanation would do here? Oh, and as far as the check for `TTY` goes, I'm actually not sure what to do here. I'm not sure whether the handle returned by `GetStdHandle` will be treated like a `TTY`. I guess we should investigate further, but I'd vouch for leaving a TODO comment in and do it in a subsequent PR?

Reply:
OK, after careful analysis, I think we can remove the check and _always_ santise the output to `stdout` since `is_tty` will always resolve to true for `Stdio` handle type. However, I think we need to keep the check in case of `OsOther` handle type which can encapsulate user-specified handle to redirected stdio. I'm thinking here of piping and mainly on Unix, but perhaps it might also be possible on Windows.

I've done that now in f76e7c4.

Reply:
Thanks, I understand this a lot better now. Thanks for the explanation and adding it to the docs!